src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -86,11 +86,145 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
                 // being overly careful to only try to format syntax forms they care about.
                 TryFormatCSharpCodeBlock(context, edits, source, node);
                 TryFormatSingleLineDirective(context, edits, source, node);
+                TryFormatExplicitCodeBlock(context, edits, source, node);
             }
 
             return edits;
         }
 
+        private static void TryFormatExplicitCodeBlock(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            SyntaxNode openBraceNode;
+            SyntaxNode codeNode;
+            SyntaxNode closeBraceNode;
+            var additionalIndentation = false;
+
+            // TODO: Split the if's into their own methods
+
+            // complex situations like
+            // @{
+            //  void Method(){
+            //      @(DateTime.Now)
+            //  }
+            // }
+            if (node is CSharpRazorBlockSyntax csharpRazorBlock &&
+                csharpRazorBlock.Parent is CSharpCodeBlockSyntax innerCodeBlock &&
+                csharpRazorBlock.Parent.Parent is CSharpCodeBlockSyntax outerCodeBlock)
+            {
+                var csharpRazorRange = csharpRazorBlock.GetRange(source);
+
+                codeNode = csharpRazorBlock;
+                openBraceNode = outerCodeBlock.Children.PreviousSiblingOrSelf(innerCodeBlock);
+                closeBraceNode = outerCodeBlock.Children.NextSiblingOrSelf(innerCodeBlock);
+            }
+            // void Method()
+            // {
+            //     <div></div>
+            // }
+            else if (node is MarkupBlockSyntax markupBlockNode &&
+                markupBlockNode.Parent is CSharpCodeBlockSyntax cSharpCodeBlock)
+            {
+                var previousChild = cSharpCodeBlock.Children.PreviousSiblingOrSelf(markupBlockNode);
+                var nextChild = cSharpCodeBlock.Children.NextSiblingOrSelf(markupBlockNode);
+
+                codeNode = markupBlockNode;
+                openBraceNode = previousChild;
+                closeBraceNode = nextChild;
+            }
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            else if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
+                statement.Body is CSharpStatementBodySyntax csharpStatementBody)
+            {
+                openBraceNode = csharpStatementBody.OpenBrace;
+                codeNode = csharpStatementBody.CSharpCode;
+                closeBraceNode = csharpStatementBody.CloseBrace;
+
+                additionalIndentation = true;
```

Comment:
What does this mean? If all of these do things with braces, why is this one special and gets more indentation? Perhaps a comment explaining would be good.

Reply:
So the thinking here is that the IndentationContext doesn't have the full info. Consider:
`@{Debugger.Launch();}`. We know this should become
```razor
@{
    Debugger.Launch();
}
```
but the Indentation system thinks that the Razor and Html desired indentation levels are 0 (presumably because things are sharing a line, though I don't know why the other scenarios don't ask need this too). This is certainly a gross hack, but I'm trying to avoid boiling the whole formatting/indentation system at once.

Reply:
That makes sense, and would definitely benefit from a comment. Its a little surprising that there is no check here for things being on a single line, or anything that would make it detect one form over the other.

Reply:
We check for same-line here: https://github.com/dotnet/aspnetcore-tooling/pull/4032/files#diff-90e26854fd9209b5a5868aba10c7275281a167587f6df0515448a3fea7f2b852R176.

Not sure what you mean about "detect one form over the other"?

Reply:
> We check for same-line here

👍 

> Not sure what you mean about "detect one form over the other"?

I was assuming from your comment that we only want to do this when the code snippet is `@{ blah }` and not `@{\n blah \n}`. Looking at where you've linked to I can see that is the case, in that this `additionalIndentation` variable is only used in the single line case.

I think this all just goes back to the style question of do we have one method per code snippet we're trying to find and format, or do we combine together and share as much code as possible. This is an example of the latter making things a little harder to follow. A comment will help though.